keys: from_nodetool_style_string don't split single partition keys

tarasveretilnyk opened the pull request:
Users with single-column partition keys that contain colon characters
were unable to use certain REST APIs and 'nodetool' commands, because the
API split key by colon regardless of the partition key schema.

Affected commands:
- 'nodetool getendpoints'
- 'nodetool getsstables'
Affected endpoints:
- '/column_family/sstables/by_key'
- '/storage_service/natural_endpoints'

Refs: https://github.com/scylladb/scylladb/issues/16596 - This does not fully fix the issue, as users with compound
keys will face the issue if any column of the partition key contains
a colon character.

Backport: required, it is a bug fix

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18114/) | :white_check_mark: | sif |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18114/) | :white_check_mark: | spider8.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18114/) | :white_check_mark: | spider8.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18114/)<br> :small_blue_diamond: test/boost/keys_test.cc<br> :small_blue_diamond: rest_api/test_storage_service | :white_check_mark: | spider8.cloudius-systems.com |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18114/) | :white_check_mark: | spider5.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18114/) | :white_check_mark: | spider5.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18114/)<br> :small_blue_diamond: test/boost/keys_test.cc<br> :small_blue_diamond: rest_api/test_storage_service | :white_check_mark: | spider5.cloudius-systems.com |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18114/) | :white_check_mark: | spider1.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18114/) | :white_check_mark: | spider1.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18114/)<br> :small_blue_diamond: test/boost/keys_test.cc<br> :small_blue_diamond: rest_api/test_storage_service | :white_check_mark: | spider1.cloudius-systems.com |
| dtest | [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5895/) | :white_check_mark: |  |
|  | [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/17215/) | :white_check_mark: |  |
|  | [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/9122/) | :white_check_mark: |  |

#### Build Details:
- Duration: 6 hr 1 min

tarasveretilnyk commented on test/rest_api/test_storage_service.py:
```diff
@@ -629,6 +629,24 @@ def test_storage_service_get_natural_endpoints(cql, rest_api, tablets_enabled, s
 
             assert resp.json() == [rest_api.host]
 
+@pytest.mark.parametrize("tablets_enabled", ["true", "false"])
```
Done

avikivity commented on api/api-doc/storage_service.json:
```diff
@@ -589,6 +589,14 @@
+                     "required":false,
+                     "allowMultiple":false,
+                     "type":"string",
+                     "paramType":"query"
```
I think it's wrong to add the delimiter to the API.

Better introduce a new endpoint that accepts a list of components, and do the splitting in nodetool.

mykaul replied:
isn't it an overkill? How will users of the (old) endpoint (Manager?) know which one to use?

denesb replied:
On the one hand I also think a new endpoint is overkill for this. We got away for years without fixing this issue -- which is just to say that keys which need custom delimiters are very rare.
On the other hand, how will a non-human client like manager choose the right delimiter? Try non-alpha characters until one is found which is not in the key? What if all are in the key? An API which just accepts key components separated is just cleaner.

An alternative approach to fix is to allow escaping instances of the separator in the key.

tarasveretilnyk replied:
I think that adding new endpoint that accepts list query parameter might work.
There  are two common ways to pass an array as query parameter:
1.  Repeating the key for each value(&key=part1&key=part2)
2. Comma separated values(key=part1,part2,part3)
Where part1,part2,...,partn are url encoded, so it solves delimiter issues.

However, from what I can see, seastar doesn't support array query parameters(they are just stored as std::unordered_map<sstring, sstring>). So it will probably requires changes to  seastar library.

denesb replied:
If we could make array parameter work, we wouldn't need a new API endpoint.

nyh replied:
@tarasveretilnyk I mentioned this in another thread, but I'll repeat it again here: The "common way 2" you mentioned is wrong, and based on a **misconception** of who does URL encoding when. The entire value, `part1,part2,part3` is URL-encoded. That is what the standard says. If you want to _additionally_ URL-encode each inner part separately (part1 et al.) , this double-application of the URL encoding and is not part of any standard or common approach. In fact, if you want to have a second encoding for the inner parts (part1 et al.) there is no standard reason why this second encoding must be URL encoding (it's even confusing if it's URL encoding) and it can be something else - e.g., backslashing commas and backslashes.

The first approach you mentioned - repeating the key - is indeed a very valid approach, I just don't know if Seastar supports it (probably not - https://github.com/scylladb/seastar/issues/693), and since we used some sort of JSON api/api-docs to "define" our API (CC @amnonh) I have no idea if that format actually supports this approach of specifying the same parameter key multiple times to mean a list. Would be nice if it did - I think that would be the best approach. This is also the approach I suggested in hhttps://github.com/scylladb/scylladb/issues/16596#issuecomment-1873327472- but apparently it won't work without fixing Seastar (https://github.com/scylladb/seastar/issues/693). But I think it's worth fixing.

denesb commented on api/storage_service.cc:
```diff
@@ -721,7 +721,7 @@ static
 rest_get_natural_endpoints(http_context& ctx, sharded<service::storage_service>& ss, const_req req) {
         auto keyspace = validate_keyspace(ctx, req);
-        auto res = ss.local().get_natural_endpoints(keyspace, req.get_query_param("cf"), req.get_query_param("key"));
+        auto res = ss.local().get_natural_endpoints(keyspace, req.get_query_param("cf"), req.get_query_param("key"), req.get_query_param("key_delimiter"));
```
This makes the key_delimiter a de-facto required parameter, breaking all current clients of the API.
If the parameter is not required, you need to handle the case where it is not provided.

tarasveretilnyk commented:
v4: Rebased and removed fix for compound keys
Seems like customers occasionally face those issue, so I suggest addressing fix for compound keys in a separate PR, as it anyway depends on a Seastar patch. I also think the smaller fix for single partition keys should be backported.

@denesb @avikivity, please re-review when you have a moment.
